aggregator/deduplication.md

```diff
@@ -0,0 +1,107 @@
+# Deduplicating Used Concepts
+
+## Why?
+Although the database is perfectly capable of handling requests to add duplicate data, 
+the scale of duplication in the full set of used concepts means that it is prudent 
+to filter duplicates out in the aggregator, rather than allowing the database to 
+deal with it.
+
+In the 2022-08-22 snapshot, there were roughly 3.7 Million concepts, but 
+fewer than a quarter of a million distinct concepts.
+
+The overhead in rendering each of these into a bulk request, and posting
+them all to Elasticsearch is much greater than that of removing duplicates.
+
+## Time Trials
+
+Running on a macbook using docker compose, and a local copy of the snapshot, thus:
+`cat works.json | docker compose run -T aggregator`, I trialled a few approaches.
+
+In each, the Elasticsearch database was already fully populated, so all "updates"
+would return "noop". Elasticsearch had ben left running with `docker compose run elasticsearch`
+
+Simply sending all the data to Elasticsearch took over six minutes.
+Various methods of deduplication consistently took below four minuts.
+
+## Why do it this way?
+
+I tried a number of different deduplication approaches, settling on the statefulMapConcat approach.
+
+The full snapshot will not be run frequently, but it still needs to run within sensible 
+bounds of time and resource usage.
+
+### Baselines
+
+The chosen approach had to beat these two naive baselines:
+
+1 - letting the database deal with it (see above)
+2 - gathering all the concepts in a list and calling distinctBy 
+    (memory peaked at 1.5GiB, CPU at 300%, took about 4 minutes)
+
+### statefulMapConcat
+First was a Flow using statefulMapConcat and a mutable Set.
+```
+   def deduplicateFlow: Flow[UsedConcept, UsedConcept, NotUsed] = 
+    Flow[UsedConcept].statefulMapConcat { () =>
+    val seen: MutableSet[String] = MutableSet.empty[String];
```

Comment:
Didn't have brain space to think about this yesterday, but a naïve suggestion for slightly reducing memory usage would be instead to use a `MutableSet[Int]` for storing `concept.identifier.hashCode`s, which I would expect to be smaller than the strings. The other _distressingly nerdy_ approach that occurs to me would be to use a Bloom filter(!)

Reply:
Good point!  I did have a thought along those lines, which I then dismissed with the thought "It's doing hashes behind the scenes anyway". However, that was a silly thought.

Reply:
In fact at that point I suppose you may as well hash the whole concept rather than concerning yourself with its ID...

Reply:
the id is the important thing.  Because we are storing the label, it is possible for two UsedConcept objects with the same ID to be different.

Reply:
Nice.  Peaks at 510MiB now (580 for the current snapshot)